include/navigation/infomax.h

```diff
@@ -90,7 +133,21 @@ class InfoMax : public VisualNavigationBase
         const auto id = MatrixType::Identity(m_Weights.rows(), m_Weights.rows());
         const auto sumYU = (m_Y.array() + m_U.array()).matrix();
         const FloatType learnRate = m_LearningRate / (FloatType) m_U.rows();
+
         m_Weights.array() += (learnRate * (id - sumYU * m_U.transpose()) * m_Weights).array();
+
+        /*
+         * If the learning rate is too high, we may end up with NaNs in our
+         * weight matrix, which will silently muck up subsequent calculations.
+         * I don't *think* this will ever be an issue with a sensibly small
+         * learning rate, but if so, the learning rate could be reduced at this
+         * point instead of just bailing out.
+         */
+        const auto ptr = m_Weights.data();
+        const auto check = [](auto val) { return std::isnan(val); };
+        if (std::any_of(&ptr[0], &ptr[m_Weights.size()], check)) {
+            throw WeightsBlewUpError{};
+        }
```

Comment:
By randomly googling, I saw this nan check:
```c++
return ((x.array() == x.array())).all();
```
which should get vectorized/parallelised by Eigen so likely to be faster

Reply:
Ah this does look nicer than my hack. Thanks!

Reply:
No worries - I dunno about nicer but perhaps idiomatic Eigen rather than idiomatic C++